rfcs/0012-compaction-state-persistence.md

```diff
@@ -0,0 +1,487 @@
+
+**Rationale**: Reuses proven patterns for atomic updates, version checking, and conflict resolution that are already battle-tested in SlateDB.
+
+#### **4. Recovery Strategy**
+**Decision**: Smart recovery based on work preservation efficiency:
+- **High efficiency** (>80% work preserved): Resume from last completed output SST
+- **Low efficiency**: Full retry with exponential backoff
+
+**Rationale**: Balances complexity vs. value - don't over-optimize recovery for cases where restart is simpler.
+
+### **Data Model**
+```
+Compaction (1) ──→ (N) CompactionJob
+    ├── sources: Vec<SourceId>
+    ├── destination: u32  
+    ├── job_attempts: Vec<CompactionJob>
+    └── current_phase: CompactionPhase
+
+CompactionJob
+    ├── attempt_number: u32
+    ├── progress: CompactionProgress  
+    ├── status: CompactionJobStatus
+    └── existing_output_ssts: Option<Vec<SsTableId>>
+
+CompactionProgress
+    ├── input_ssts_completed: Vec<SsTableId>
+    ├── output_ssts_written: Vec<SsTableId>
+    ├── bytes_read/written: u64
+    └── current_phase: CompactionPhase
+```
+
+### **Persistent State Storage**
+
+#### **Object Store Layout**
+The compaction state is persisted to the object store following the same CAS pattern as manifests, ensuring consistency and reliability:
+
+```
+/compactor_state_000000001.fbs  # First compactor state
+/compactor_state_000000002.fbs  # Updated state after compactions
+/compactor_state_000000003.fbs  # Current state
+```
+
+#### **CompactionState Structure**
+The persistent state contains the complete view of all compaction activity:
+
+```rust
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CompactionState {
+    /// Fencing token to ensure single active compactor
+    /// Incremented each time a new compactor takes control
+    pub compactor_epoch: u64,
+    
+    /// Next state ID to use for CAS operations
+    /// Current state ID = this value - 1
+    pub next_compactor_state_id: u64,
+    
+    /// All currently active compactions indexed by ID
+    /// Includes queued, running, and recently completed compactions
+    pub active_compactions: HashMap<CompactionId, Compaction>,
+    
+    /// Historical record of completed compaction jobs
+    /// Used for statistics and debugging, cleaned up by GC
+    pub job_history: HashMap<CompactionJobId, CompactionJob>,
+    
+    /// Queue of pending manual compaction requests
+    /// Processed in priority order by the active compactor
+    pub manual_compaction_queue: Vec<ManualCompactionRequest>,
+    
+    /// Aggregated performance and reliability statistics
+    pub statistics: CompactionStatistics,
+    
+    /// Timestamp of last coordination with garbage collector
+    /// Used to prevent GC from removing SSTs still being compacted
+    pub last_gc_coordination_ts: DateTime<Utc>,
+    
+    /// Active configuration settings for compaction behavior
+    pub configuration: CompactionConfiguration,
```

Comment:
Curious about this as well. As with the stats, the `Compactor` takes options (config) right now. This is the pattern we typically have.